src/Microsoft.Data/PrimitiveColumnContainer.cs

```diff
@@ -69,91 +84,224 @@ public PrimitiveColumnContainer(long length = 0)
+        public void Append(T? value)
         {
             if (Buffers.Count == 0)
             {
                 Buffers.Add(new DataFrameBuffer<T>());
+                NullBitMapBuffers.Add(new DataFrameBuffer<byte>());
             }
             DataFrameBuffer<T> lastBuffer = Buffers[Buffers.Count - 1];
             if (lastBuffer.Length == lastBuffer.MaxCapacity)
             {
                 lastBuffer = new DataFrameBuffer<T>();
+                Buffers.Add(lastBuffer);
+                NullBitMapBuffers.Add(new DataFrameBuffer<byte>());
             }
-            lastBuffer.Append(value);
+            lastBuffer.Append(value ?? default);
+            SetValidityBit(Length, value.HasValue ? true : false);
             Length++;
         }
 
+        public void AppendMany(T? value, long count)
+        {
+            if (!value.HasValue)
+            {
+                NullCount += count;
+            }
+            while (count > 0)
+            {
+                if (Buffers.Count == 0)
+                {
+                    Buffers.Add(new DataFrameBuffer<T>());
+                    NullBitMapBuffers.Add(new DataFrameBuffer<byte>());
+                }
+                DataFrameBuffer<T> lastBuffer = Buffers[Buffers.Count - 1];
+                if (lastBuffer.Length == lastBuffer.MaxCapacity)
+                {
+                    lastBuffer = new DataFrameBuffer<T>();
+                    Buffers.Add(lastBuffer);
+                    NullBitMapBuffers.Add(new DataFrameBuffer<byte>());
+                }
+                int allocatable = (int)Math.Min(count, lastBuffer.MaxCapacity);
+                lastBuffer.EnsureCapacity(allocatable);
+                lastBuffer.Span.Slice(lastBuffer.Length, allocatable).Fill(value ?? default);
+                lastBuffer.Length += allocatable;
+                Length += allocatable;
+
+                DataFrameBuffer<byte> lastNullBitMapBuffer = NullBitMapBuffers[NullBitMapBuffers.Count - 1];
+                int nullBitMapAllocatable = (int)Math.Ceiling(allocatable / 8.0);
+                lastNullBitMapBuffer.EnsureCapacity(nullBitMapAllocatable);
+                _modifyNullCountWhileIndexing = false;
+                for (long i = Length - count; i < Length; i++)
+                {
+                    SetValidityBit(i, value.HasValue ? true : false);
+                }
+                _modifyNullCountWhileIndexing = true;
+                lastNullBitMapBuffer.Length += nullBitMapAllocatable;
+                count -= allocatable;
+            }
+        }
+
+        public bool IsValid(long index) => NullCount == 0 || GetValidityBit(index);
+
+        /// <summary>
+        /// A null value has an unset bit
+        /// A NON-null value has a set bit
+        /// </summary>
+        /// <param name="index"></param>
+        /// <param name="value"></param>
+        private void SetValidityBit(long index, bool value)
+        {
+            if ((uint)index > Length)
```

Comment:
```suggestion
            if ((ulong)index > Length)
```
or is it safe with `uint`?

Reply:
Good catch. It should be ulong. Copy-paste error